packages/opal-common/opal_common/synchronization/hierarchical_lock.py

```diff
@@ -0,0 +1,85 @@
+import asyncio
+from contextlib import asynccontextmanager
+from typing import Set
+
+
+class HierarchicalLock:
+    """A hierarchical lock for asyncio.
+
+    - If a path is locked, no ancestor or descendant path can be locked.
+    - Conversely, if a child path is locked, the parent path cannot be locked
+      until all child paths are released.
+    """
+
+    def __init__(self):
+        # locked_paths: set of currently locked string paths
+        self._locked_paths: Set[str] = set()
+        # Map of tasks to their acquired locks for re-entrant protection
+        self._task_locks: dict[asyncio.Task, Set[str]] = {}
+        # Internal lock for synchronizing access to locked_paths
+        self._lock = asyncio.Lock()
+        # Condition to wake up tasks when a path is released
+        self._cond = asyncio.Condition(self._lock)
+
+    @staticmethod
+    def _is_conflicting(p1: str, p2: str) -> bool:
+        """Check if two paths conflict with each other."""
+        return p1 == p2 or p1.startswith(p2) or p2.startswith(p1)
+
+    async def acquire(self, path: str):
+        """Acquire the lock for the given hierarchical path.
+
+        If an ancestor or descendant path is locked, this will wait
+        until it is released.
+        """
+        task = asyncio.current_task()
+        if task is None:
+            raise RuntimeError("acquire() must be called from within a task.")
+
+        async with self._lock:
+            # Prevent re-entrant locking by the same task
+            if path in self._task_locks.get(task, set()):
+                raise RuntimeError(f"Task {task} cannot re-acquire lock on '{path}'.")
+
+            # Wait until there is no conflict with existing locked paths
+            while any(self._is_conflicting(path, lp) for lp in self._locked_paths):
```

Comment:
Correct me if I'm wrong but wouldn't this cause you to aquire the condition underlying lock ( which is a single lock for all the tasks ), and then waiting for the path to be unlocked, which blocks also other paths from being processed ( because the lock is still locked )